src/ray/core_worker/transport/dependency_resolver.h

```diff
@@ -39,19 +44,20 @@ class LocalDependencyResolver {
   ///
   /// Postcondition: all direct call id arguments that haven't been spilled to plasma
   /// are converted to values and all remaining arguments are arguments in the task spec.
-  void ResolveDependencies(TaskSpecification &task, std::function<void()> on_complete);
+  void ResolveDependencies(TaskSpecification &task, std::function<void(Status)> on_complete);
 
   /// Return the number of tasks pending dependency resolution.
   /// TODO(ekl) this should be exposed in worker stats.
   int NumPendingTasks() const { return num_pending_; }
 
  private:
   /// The in-memory store.
-  std::shared_ptr<CoreWorkerMemoryStore> in_memory_store_;
+  CoreWorkerMemoryStore& in_memory_store_;
```

Comment:
Why not keep shared_ptr? Ray references have caused many destruction order issues.

Reply:
I believe it gives a better understanding of ownership. Using T* or T& means that this class doesn't own this object. 

Destruction ordering issues should just be fixed. shared_ptr should only be used when sharing data across threads.

Some reference [here](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f7-for-general-use-take-t-or-t-arguments-rather-than-smart-pointers) and [here](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rr-unique)

Reply:
How would you fix it for lambdas? I commonly see this pattern:

```
post([this]() {
   this->some_ref->something()
 });
```

This will generate a segfault if the callback is executed after the destruction of `this`. If the callback has a shared_ptr to this (shared_from_this()) or captures shared ptrs to `some_ref`, then this becomes much safer.

Reply:
If `this` is shared across threads (like post to another thread), we can make this shared_from_this, but the member variable inside this doesn't need to be shared_ptr. It explicitly specifies the ownership.

If this is not shared across threads, maybe we should drain the task first?